addons/test_mass_mailing/tests/test_mailing.py

```diff
@@ -420,19 +420,21 @@ def test_mailing_w_blacklist(self):
         with self.mock_mail_gateway(mail_unlink_sent=False):
             mailing.action_send_mail()
 
-        self.assertMailTraces(
-            [{
+        expected_traces = [{
                 'email': record.email_normalized,
                 'email_to_mail': record.email_from,
             } for record in recipients[:3]] + [{
                 'email': record.email_normalized,
                 'email_to_mail': record.email_from,
                 'failure_type': 'mail_bl',
                 'trace_status': 'cancel',
-            } for record in recipients[3:]],
-            mailing, recipients, check_mail=True
-        )
+            } for record in recipients[3:]]
+        self.assertMailTraces(expected_traces, mailing, recipients, check_mail=True)
         self.assertEqual(mailing.canceled, 2)
+        self.assertEqual(len(self.env['mail.mail'].sudo().search([('mailing_id', '=', mailing.id)])), 3,
+                         "Only the 3 sent mails have been created, the canceled ones have not been created")
+        expected_message_count = [0 if t.get('trace_status') == 'cancel' else 1 for t in expected_traces]
+        self.assertMessageRecordsCount(recipients, expected_message_count)
```

Comment:
Strange, would probably include it in assertMailTraces: a canceled trace (maybe with a custom additional keyword ?) would check that the trace is not linked to any mail.mail / mail.message. That way it is managed directly by the core tool and we don't need custom check. Would that be possible ?